### ISSUE TITLE
About bpr loss implementation and regularization

### ISSUE THREAD

**Issue by keeleyadam** [lyst/lightfm#302]
Hi @maciejkula , 
LightFM helps me a lot in my learning recommendation algorithms, thank you. I am very interested in the implementation of bpr and warp losses in lightfm.
I read the BPR paper [https://arxiv.org/pdf/1205.2618.pdf ](url), in which they maximize BPR-OPT.
![github1](https://user-images.githubusercontent.com/34033376/40474802-8ec657f6-5f72-11e8-85fa-313c0af9a0ec.PNG)
Use SGD to minimize the negative of BPR, the gradient is as following fig, equation 1.
![github4](https://user-images.githubusercontent.com/34033376/40477891-315cd164-5f7a-11e8-97a7-55987d84f296.PNG)

I studied the method warp_update in _lightfm_fast.pyx.template, and found that the gradients were calculated in the following way.
loss = 1 - sigmoid(x_uij)
1. gradient for biases of positive item: -loss
2. for biases of negative item: loss
3. for biases of user: loss
4. for embeddings of positive item: -loss * p_u
5. for embeddings of negative item: loss * p_u
6. for embedding of user: -loss * (q_i - q_j)

![github3](https://user-images.githubusercontent.com/34033376/40478101-c40abad0-5f7a-11e8-8172-4578313bb1d7.PNG)
So there are two points I don't understand:
The calculations of gradients 1, 2, 4, 5, 6 seem right to me, but isn't biases of user (b_u) subtracted by itself, as the above fig? Why gradient for biases of user equals loss.
Gradients seem to be caculated by the first part of equation 1, so it is ok to leave out the second part of equation 1? 
Maybe I missed something important, I would appreciate it very much if you could give me some help.

![github5](https://user-images.githubusercontent.com/34033376/40484436-3a538b78-5f8e-11e8-881e-a1ab0dbd3301.PNG)
And about regularization, I searched "lazy regularization" and "weight decay" in google, but couldn't find useful explaination to help me understand why `features[feature, component] *= (1.0 + alpha * local_learning_rate)` works. Are there any blog post explaining this theory, or could you please tell me keywords which I can use to look up. Thank you very much.

**Comment by eggie5** [comment]
> The calculations of gradients 1, 2, 4, 5, 6 seem right to me, but isn't biases of user (b_u) subtracted by itself, as the above fig?

Yes, that is true, for pair-wise loss, the user biases cancel out.

**Comment by JohnPaton** [comment]
Hey @maciejkula, I'm experiencing diverging models (embeddings are exploding) and while digging around I was also confused by the `lazy regularization` line. In particular I'm confused why you're scaling the biases and embeddings _up_ instead of down? Isn't the point to reduce their magnitude?

I'm definitely missing something, I just can't figure out what!

**Comment by maciejkula** [comment]
One thing I would look at is whether you have any users/items without any
features: this may cause problems. If your model continues to diverge even
with a very low learning rate this could be the culprit.

In training I scale up features I see, but when computing predictions I
scale everything down by a global regularization scale (
https://github.com/lyst/lightfm/blob/master/lightfm/_lightfm_fast.pyx.template#L311).
What this gives me is a global decay that affects mostly the features I
don't see very much.

**Comment by JohnPaton** [comment]
Ah, that makes sense. 

I'm using identity features for users and they all have at least 5 other categorical features, and my items have no identity features (for easier fold in) but do all have at least 11 categorical features.

Could the lazy regularisation be scaling certain parameters up too much such that they completely dwarf everything else? I've just observed a major performance boost by significantly reducing the amount of regularisation.

**Comment by maciejkula** [comment]
I don't think it's to do with regularization being lazy: no or very little regularization is usually optimal.

**Comment by SimonCW** [comment]
I’m closing this issue because it has been inactive for a long time. If you still encounter the problem, please open a new issue.

Thank you!
